server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java

```diff
@@ -0,0 +1,494 @@
+    }
+    return false;
+  }
+
+  /**
+   * Create initial system props for the instance. If the node already exists, no action is
+   * performed.
+   *
+   * @param context
+   *          the server context.
+   * @param initProps
+   *          map of k, v pairs of initial properties.
+   */
+  public synchronized static void initSysProps(final ServerContext context,
+      final Map<String,String> initProps) {
+    PropCacheKey sysPropKey = PropCacheKey.forSystem(context.getInstanceID());
+    createInitialProps(context, sysPropKey, initProps);
+  }
+
+  /**
+   * Create initial properties if they do not exist. If the node exists, initialization will be
+   * skipped.
+   *
+   * @param context
+   *          the system context
+   * @param propCacheKey
+   *          a prop id
+   * @param props
+   *          initial properties
+   */
+  public static void createInitialProps(final ServerContext context,
+      final PropCacheKey propCacheKey, Map<String,String> props) {
+
+    try {
+      ZooReaderWriter zrw = context.getZooReaderWriter();
+      if (zrw.exists(propCacheKey.getPath())) {
+        return;
+      }
+      VersionedProperties vProps = new VersionedProperties(props);
+      zrw.putPersistentData(propCacheKey.getPath(), codec.toBytes(vProps),
+          ZooUtil.NodeExistsPolicy.FAIL);
+    } catch (InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      throw new PropStoreException("Interrupted creating node " + propCacheKey, ex);
+    } catch (Exception ex) {
+      throw new PropStoreException("Failed to create node " + propCacheKey, ex);
+    }
+  }
+
+  public PropStoreMetrics getMetrics() {
+    return cacheMetrics;
+  }
+
+  @Override
+  public void create(PropCacheKey propCacheKey, Map<String,String> props) {
+
+    try {
+      VersionedProperties vProps = new VersionedProperties(props);
+      String path = propCacheKey.getPath();
+      zrw.putPersistentData(path, codec.toBytes(vProps), ZooUtil.NodeExistsPolicy.FAIL);
+    } catch (IOException | KeeperException | InterruptedException ex) {
+      throw new PropStoreException("Failed to serialize properties for " + propCacheKey, ex);
+    }
+  }
+
+  /**
+   * get or create properties from the store. If the property node does not exist in ZooKeeper,
+   * legacy properties exist, they will be converted to the new storage form and naming convention.
+   * The legacy properties are deleted once the new node format is written.
+   *
+   * @param propCacheKey
+   *          the prop cache key
+   * @return The versioned properties or null if the properties do not exist for the id.
+   * @throws PropStoreException
+   *           if the updates fails because of an underlying store exception
+   */
+  @Override
+  public @NonNull VersionedProperties get(final PropCacheKey propCacheKey)
+      throws PropStoreException {
+    checkZkConnection(); // if ZK not connected, block, do not just return a cached value.
```

Comment:
What scenario does this protect against?  Immediately after this method returns before it executes any subsequent statements, ZK could become disconnected.

Reply:
On a disconnect, the cache will be cleared and the event propagated. This check the guaranties that all subsequent reads block in case ZooKeeper recovers. Any code that passes this check and then ZooKeeper disconnects will either be retrying (reties with zoo reader writer) or have the ZooKeeper calls fail.

Reply:
It seems like this code adds complexity without delivering a reliable benefit because there is a race condition.  The code after this that retires in the case failure is where the buck stops and its what always has to work reliably.  This code can make it harder to test the code that retries in the case of failure. Also if this code is not working we may not ever know because the code that retries in the case of failure picks up the slack.  It seems like this check could be removed to simplify the code and make it easier to test faults in the retry code.

Reply:
Again - this code was to handle the case when ZooKeepeer sessions errors would be transient. Primarily the case of a ZooKeeper node failure, but when there is still a quorum.  When that happens, ZooKeeper sends a disconnected event and then reconnected event, if the connection established before time-out. In those cases, you are in an unknown state while disconnected - you have not failed, but you are also unable to establish the current state, so you likel;y should not proceed as if everything is okay either.

The check / ready monitor will block until the connection event is received - future reads may succeed, but if you allow them to try without the connection they will fail.  So the code is making a best effort to handle that situation. Because of the asynchronous nature and the reality that failures can happen anywhere, I don't think that the race can be completely eliminated, but the errors are handled and the code should allow for a more gracefully handling of disconnect events than just allowing the errors to occur.

Transient ZooKeeper errors are not uncommon, and this would should allow better handling of things like a rolling ZooKeeper restart. If that is not handled 100% in the current code, this still allows us to move in that direction and then add additional handling that where it is needed.

I can remove the code if you disagree - or maybe I've missed the point and we are talking past each other?